pkg/controller/source/channel_source.go

```diff
@@ -0,0 +1,293 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package source
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/buffer"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/kubernetes-sigs/controller-runtime/pkg/controller/event"
+	"github.com/kubernetes-sigs/controller-runtime/pkg/controller/eventhandler"
+	"github.com/kubernetes-sigs/controller-runtime/pkg/controller/source/internal"
+	"github.com/kubernetes-sigs/controller-runtime/pkg/runtime/inject"
+
+	logf "github.com/kubernetes-sigs/controller-runtime/pkg/runtime/log"
+)
+
+const (
+	// initialBufferSize is the initial number of event notifications that can be buffered.
+	initialBufferSize = 1024
+)
+
+var csLog = logf.KBLog.WithName("source").WithName("ChannelSource")
+
+// ChannelSource is used to provide a source of events originating outside the cluster
+// (eh.g. GitHub Webhook callback).  ChannelSource requires the user to wire the external
+// source (eh.g. http handler) to write GenericEvents to the underlying channel.
+type ChannelSource struct {
+	informer *ChannelSourceInformer
```

Comment:
Why do we need an informer for this?  I envisioned reading events from the Channel and passing them directly to the Event handler